### PR TITLE
store, daemon, client, cmd/snap: bubble TWOFACTOR_FAILURE up from store all the way to snap

### PR DESCRIPTION
**Pull request by chipaca**
includes a rather weird construct in `cmd/snap` that i'm not too happy about, but at least it's DRY.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -211,6 +211,21 @@ func (e *Error) Error() string {
 	return e.Message
 }
 
+// these should be self-explanatory; otherwise, let us know!
```
Comment (niemeyer) [inline client/client.go]: That comment should be implied over every non-documented interface. :-)  Can we please drop it?

File: cmd/snap/cmd_login.go [hunk 2]
```diff
@@ -54,28 +53,33 @@ func init() {
 }
 
 func requestLoginWith2faRetry(username, password string) error {
+	var otp []byte
+	var err error
+
+	var msgs = [3]string{
+		i18n.G("Two-factor code: "),
+		i18n.G("Bad code. Try again: "),
+		i18n.G("Wrong again. Last chance: "),
```
Comment (niemeyer) [inline cmd/snap/cmd_login.go]: s/Last chance/Once more/?

The "Last chance" makes it feel a bit like a credit card that will lock you out after three attempts.

File: cmd/snap/cmd_login.go [hunk 3]
```diff
@@ -54,28 +53,33 @@ func init() {
...
+		// first try without otp
+		_, err = cli.Login(username, password, string(otp))
+		if i > 2 || !client.IsTwoFactorError(err) {
 			return err
 		}
 
-		fmt.Print(i18n.G("Two-factor code: "))
-		reader := bufio.NewReader(os.Stdin)
+		reader.Reset(os.Stdin)
+		fmt.Print(msgs[i])
```
Comment (niemeyer) [inline cmd/snap/cmd_login.go]: Should this use the package-level Stdout, so it's redirected for testing?

I know we have a few tests which are currently missing this and `go test` ends up breaking a few lines up already on test runs.

File: daemon/api.go [hunk 4]
```diff
@@ -196,19 +196,30 @@ func loginUser(c *Command, r *http.Request) Response {
...
-	}
-	if err != nil {
+	case store.Err2faFailed:
+		return SyncResponse(&resp{
+			Type: ResponseTypeError,
+			Result: &errorResult{
+				Kind:    errorKindTwoFactorFailed,
+				Message: err.Error(),
+			},
+			Status: http.StatusForbidden,
```
Comment (niemeyer) [inline daemon/api.go]: Why is the status different for those two cases, when both of them fail for basically the same reason?

Also, can these blocks please be refactored so they look like each other?

File: store/auth_test.go [hunk 5]
```diff
@@ -49,6 +49,15 @@ const mockStoreNeeds2fa = `
 }
 `
 
+const mockStore2faFailedHTTPCode = 403
+const mockStore2faFailedResponse = `
+{
+    "meessage": "The provided 2-factor key is not recognised.",
```
Comment (niemeyer) [inline store/auth_test.go]: Tipo in the key. Something tells me the message is not very relevant. ;-)

**Comment by niemeyer** [comment]
A few comments, but nothing worth blocking the branch for. LGTM

File: daemon/api.go (hunk 4 above)
Comment (chipaca) [inline daemon/api.go]: 401 is basically "(more) credentials needed" whereas 403 is "bad credentials", which is the difference between these two cases

File: store/auth_test.go (hunk 5 above)
Comment (chipaca) [inline store/auth_test.go]: Good catch. And yes, message is there for completeness but we don't currently use it for anything

File: cmd/snap/cmd_login.go (hunk 2 above)
Comment (chipaca) [inline cmd/snap/cmd_login.go]: MBUAHAAA

Oh alright then

File: cmd/snap/cmd_login.go (hunk 3 above)
Comment (chipaca) [inline cmd/snap/cmd_login.go]: Ok

File: client/client.go (hunk 1 above)
Comment (chipaca) [inline client/client.go]: Yup

File: client/client.go [hunk 6]
```diff
@@ -211,6 +211,20 @@ func (e *Error) Error() string {
 	return e.Message
 }
 
+const (
+	ErrorKindTwoFactorRequired = "two-factor-required"
+	ErrorKindTwoFactorFailed   = "two-factor-failed"
+)
+
+func IsTwoFactorError(err error) bool {
```
Comment (zyga) [inline client/client.go]: Not a big deal but a ```// comment``` would be nice, perhaps:

// IsTwoFactorError checks if the error is related to two factor authentication.

File: cmd/snap/cmd_login.go [hunk 7]
```diff
@@ -54,35 +52,40 @@ func init() {
 }
 
 func requestLoginWith2faRetry(username, password string) error {
+	var otp []byte
+	var err error
+
+	var msgs = [3]string{
```
Comment (zyga) [inline cmd/snap/cmd_login.go]: Heh, nice :-)

File: cmd/snap/cmd_login.go [hunk 8]
```diff
@@ -54,35 +52,40 @@ func init() {
...
-	_, err := cli.Login(username, password, "")
-	if err != nil {
-		// check if we need 2fa
-		if err, ok := err.(*client.Error); !ok || err.Kind != store.TwoFactorErrKind {
+	reader := bufio.NewReader(nil)
+
+	for i := 0; ; i++ {
+		// first try without otp
+		_, err = cli.Login(username, password, string(otp))
+		if i > 2 || !client.IsTwoFactorError(err) {
```
Comment (zyga) [inline cmd/snap/cmd_login.go]: Perhaps use ```len(msgs)``` so that someone changing the code above won't have to patch this

File: daemon/api.go [hunk 9]
```diff
@@ -196,19 +196,30 @@ func loginUser(c *Command, r *http.Request) Response {
 	}
 
 	discharge, err := store.DischargeAuthCaveat(loginData.Username, loginData.Password, macaroon, loginData.Otp)
-	if err == store.ErrAuthenticationNeeds2fa {
+	switch err {
+	case store.ErrAuthenticationNeeds2fa:
 		twofactorRequiredResponse := &resp{
```
Comment (pedronis) [inline daemon/api.go]: can you get rid of the var here, and just return SyncResponse like the other block (at the 3rd of these *I* would be tempted to invent some kind of helper in response.go to make these kind of responses with kind and value but oh well for now)

**Comment by niemeyer** [comment]
Looks great, thanks!

**Comment by stevenwilkin** [comment]
👍

**Pull request merged**